finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -290,35 +298,71 @@ mod tests {
             .collect()
     }
 
-    async fn keyboxes() -> Vec<KeyBox> {
-        let num_keyboxes: usize = NUM_NODES.into();
+    async fn crypto_basics() -> (Vec<(NodeIndex, AuthorityPen)>, AuthorityVerifier) {
+        let num_crypto_basics = NUM_NODES.into();
         let keystore = Arc::new(KeyStore::new());
-        let mut auth_ids = Vec::with_capacity(num_keyboxes);
-        for _ in 0..num_keyboxes {
+        let mut auth_ids = Vec::with_capacity(num_crypto_basics);
+        for _ in 0..num_crypto_basics {
             let pk = keystore.ed25519_generate_new(KEY_TYPE, None).await.unwrap();
             auth_ids.push(AuthorityId::from(pk));
         }
-        let mut result = Vec::with_capacity(num_keyboxes);
-        for i in 0..num_keyboxes {
-            result.push(KeyBox::new(
+        let mut result = Vec::with_capacity(num_crypto_basics);
+        for i in 0..num_crypto_basics {
+            result.push((
                 NodeIndex(i),
-                AuthorityVerifier::new(auth_ids.clone()),
                 AuthorityPen::new(auth_ids[i].clone(), keystore.clone())
                     .await
                     .expect("The keys should sign successfully"),
             ));
         }
-        result
+        (result, AuthorityVerifier::new(auth_ids))
     }
 
     async fn build() -> (Discovery, Vec<SessionHandler>) {
+        let crypto_basics = crypto_basics().await;
         let mut handlers = Vec::new();
-        for (keybox, address) in keyboxes().await.into_iter().zip(addresses()) {
+        for (authority_index_and_pen, address) in crypto_basics.0.into_iter().zip(addresses()) {
             handlers.push(
-                SessionHandler::new(keybox, SessionId(43), vec![address.into()])
+                SessionHandler::new(
+                    Some(authority_index_and_pen),
+                    crypto_basics.1.clone(),
+                    SessionId(43),
+                    vec![address.into()],
+                )
+                .await
+                .unwrap(),
+            );
+        }
+        (Discovery::new(Duration::from_millis(MS_COOLDOWN)), handlers)
+    }
+
+    async fn build_non_validators() -> (Discovery, Vec<SessionHandler>) {
```

Comment:
Just return the almost unique nonvalidator as the last element of the above function instead.

Reply:
Done